### PR TITLE
charity CRUD

### PR DESCRIPTION
**Pull request by zhaulanbekk**

### CODE REVIEW

File: src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    private final SubCategoryRepository subCategoryRepository;
+
+
+    public User getPrinciple() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        String email = authentication.getName();
+        return userRepository.findByEmail(email).orElseThrow(
+                () -> new NotFoundException("не найден!")
+        );
+
```
Comment (nurtilek94) [inline src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java:42]: WTF unnecessary line

File: src/main/java/kg/peaksoft/giftlistb6/db/repositories/SubCategoryRepository.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package kg.peaksoft.giftlistb6.db.repositories;
+
+import kg.peaksoft.giftlistb6.db.models.SubCategory;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface SubCategoryRepository extends JpaRepository<SubCategory, Long> {
+    @Query("select c from SubCategory c where c.name = :name")
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/db/repositories/SubCategoryRepository.java:11]: Write to the next line

File: src/main/java/kg/peaksoft/giftlistb6/dto/responses/ReservoirResponse.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import kg.peaksoft.giftlistb6.db.models.User;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@NoArgsConstructor
+public class ReservoirResponse {
+    private Long id;
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/dto/responses/ReservoirResponse.java:16]: Write to the next line

File: src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+@RestController
+@RequestMapping("/api/charities")
+@RequiredArgsConstructor
+@CrossOrigin(origins = "*", maxAge = 3600)
+@Tag(name = "Charity Api", description = "Charity CRUD operations")
+public class CharityApi {
+
+    private final CharityService charityService;
+
+    @GetMapping
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java]: Pls write @GetMapping @PostMapping @PutMapping @DeleteMapping  under @Operation annotation

File: src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    @PutMapping("{id}")
+    @Operation(summary = "Update Charity", description = "User can update charity")
+    public InnerPageCharityResponse updateCharity(@PathVariable Long id,
+                                                  @RequestBody CharityRequest request) {
+        return charityService.updateCharity(id, request);
+    }
+
+    @GetMapping("{id}")
+    @Operation(summary = "Get Charity", description = "Get Charity By Id")
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java]: Charity By Id write in lowercase

**Review by i-Coderr — COMMENTED** [review]
Commented 11.04.22 at 12.12 pm

File: src/main/java/kg/peaksoft/giftlistb6/dto/responses/UserCharityResponse.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@AllArgsConstructor
+public class UserCharityResponse {
+
+    private Long id;
+    private String fullName;
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/dto/responses/UserCharityResponse.java]: Remove fullName and add firstName, lastName

File: src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    private final CharityService charityService;
+
+    @Operation(summary = "All charities", description = "User can see all charities")
+    @GetMapping
+    public CharityResponses getAllCharities() {
+        return charityService.getAllCharityResponse();
+    }
+
+    @Operation(summary = "Save Charity", description = "User can save charities")
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java]: Charity lowercase

File: src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    @PostMapping
+    public YourCharityResponse saveCharity(@RequestBody CharityRequest request) {
+        return charityService.saveCharity(request);
+    }
+
+    @Operation(summary = "Update Charity", description = "User can update charity")
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java]: Charity lowercase

File: src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    @PutMapping("{id}")
+    public InnerPageCharityResponse updateCharity(@PathVariable Long id,
+                                                  @RequestBody CharityRequest request) {
+        return charityService.updateCharity(id, request);
+    }
+
+    @Operation(summary = "Get Charity", description = "Get charity by id")
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java]: Charity lowercase

File: src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    @GetMapping("{id}")
+    public InnerCharityResponse getCharityById(@PathVariable Long id) {
+        return charityService.getCharityById(id);
+    }
+
+    @Operation(summary = "Delete Charity", description = "User can delete own charity")
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/api/CharityApi.java]: Charity lowercase

File: src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    public User getPrinciple() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        String email = authentication.getName();
+        return userRepository.findByEmail(email).orElseThrow(
+                () -> new NotFoundException("не найден!"));
+    }
+
+    public InnerCharityResponse getCharityById(Long id) {
+        Charity charity = charityRepository.findById(id).orElseThrow(
+                () -> new NotFoundException("Charity with id: " + id + " not found!")
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java]: Write in Russian

File: src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        return new InnerPageCharityResponse(charity1.getId(), charity1.getImage(), charity1.getName(),
+                charity1.getDescription(), charity1.getCategory().getName(), charity1.getSubCategory().getName(),
+                charity1.getCondition(), charity1.getCreatedDate(), charity1.getCharityStatus(), charity1.getUser().getId(),
+                charity1.getUser().getPhoto(), charity1.getUser().getFirstName() + " " + charity1.getUser().getLastName());
+    }
+
+    public SimpleResponse deleteCharityById(Long id) {
+        User user = getPrinciple();
+        if (!user.getCharities().contains(charityRepository.findById(id).orElseThrow(() -> new NotFoundException("not found")))) {
+            throw new NotFoundException("You can't delete other charities");
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java]: Write in Russian

File: src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+                if (is) {
+                    charity.setReservoir(null);
+                } else {
+                    charity.setReservoir(user);
+                }
+                charity.setUser(charity.getUser());
+                user.setCharities(List.of(charity));
+                charity.setCharityStatus(Status.RESERVED);
+            } else
+                return new SimpleResponse("you can't reserve your charity", "");
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java]: Write in Russian

File: src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+        } else
+            return new SimpleResponse("charity is reserve", "error");
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java]: Write in Russian

File: src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+
+        return new SimpleResponse("ok", "reserved");
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java]: Write in Russian

File: src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+    private final CharityRepository charityRepository;
+    private final UserRepository userRepository;
+    private final CategoryRepository categoryRepository;
+    private final SubCategoryRepository subCategoryRepository;
+
+    public User getPrinciple() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        String email = authentication.getName();
+        return userRepository.findByEmail(email).orElseThrow(
+                () -> new NotFoundException("не найден!"));
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/db/services/CharityService.java]: Не с большой буквы

File: src/main/java/kg/peaksoft/giftlistb6/dto/responses/OtherCharityResponse.java [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+public class OtherCharityResponse {
+
+    private Long id;
+    private String image;
+    private String name;
+    private String condition;
+    private LocalDate addedDate;
+    private Status status;
+    private Long userId;
+    private String fullName;
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/dto/responses/OtherCharityResponse.java]: Remove fullName and add firstName, lastName

File: src/main/java/kg/peaksoft/giftlistb6/dto/responses/InnerPageCharityResponse.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    private String name;
+    private String description;
+    private String category;
+    private String subCategory;
+    private String condition;
+    private LocalDate addedTime;
+    private Status status;
+    private Long userId;
+    private String userImage;
+    private String fullName;
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/dto/responses/InnerPageCharityResponse.java]: Remove fullName and add firstName, lastName

File: src/main/java/kg/peaksoft/giftlistb6/dto/responses/ReservoirResponse.java [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.Setter;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class ReservoirResponse {
+
+    private Long id;
+    private String photo;
```
Comment (i-Coderr) [inline src/main/java/kg/peaksoft/giftlistb6/dto/responses/ReservoirResponse.java]: Rename to image

**Review by i-Coderr — COMMENTED** [review]
Commented 08.11.22 at 2 pm

**Comment by i-Coderr** [comment]
Make sure to resolve the conversation after fixing your mistakes

**Pull request merged**